File: 2020/04-passport-processing/04.mjs
import { promises as fs } from "node:fs"

const file = await fs.readFile("input.txt")

const required = ["byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid"]

const between = (value, from, to) = value >= from && value <= to

function validField([key, value]) {
    switch (key) {
    case "byr":
        return value.length === 4 && between(parseInt(value), 1920, 2002)
    case "iyr":
        return value.length === 4 && between(parseInt(value), 2010, 2020)
    case "eyr":
        return value.length === 4 && between(parseInt(value), 2020, 2030)
    case "hgt":
        const clipped = value.substr(0, value.length-2)
        if (value.endsWith("cm")) {
            return between(parseInt(clipped), 150, 193))
        } 
        return value.endsWith("in") && between(parseInt(clipped), 59, 76)
    case "hcl":
        return value.startsWith("#") && 
    }
}

for (const passport of file.split("\n\n")) {
    const fields = passport.split(/ |\n/).map(str => str.split(":"))
    const hasRequired = required.every(([key, _] => required.includes(key))
    const validFields
}

File: 2020/04-passport-processing/04.hs
import Data.List.Split
import Data.Text (isSuffixOf, pack)
import Data.Char
import Debug.Trace

main = readFile "input.txt" >>= print . length . filter validPassport . splitOn "\n\n"

required = ["byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid"]

fields = map (fieldPair) . words

fieldPair str = (key, value)
    where [key, value] = splitOn ":" str

validField "byr" value = validYear 1920 2002 value
validField "iyr" value = validYear 2010 2020 value
validField "eyr" value = validYear 2020 2030 value
validField "hgt" value = validHeight "cm" 150 193 value || validHeight "in" 59 76 value
validField "hcl" ('#':color) = length color == 6 && all isHexDigit color
validField "ecl" value = value `elem` eyeColors
validField "pid" value = length value == 9 && all isDigit value
validField "cid" _ = True
validField _ _ = False

validHeight prefix from to value = 
    (pack prefix) `isSuffixOf` (pack value) && num >= from && num <= to
    where num = read $ take (length value - length prefix) value
validYear from to year = length year == 4 && num >= from && num <= to
    where num = read year

eyeColors = ["amb", "blu", "brn", "gry", "grn", "hzl", "oth"]

validPassport str = all (`elem` (map fst passport)) required && all (uncurry validField) passport
    where passport = fields str

